packages/signed-cid/src/eip712Codec.ts

```diff
@@ -0,0 +1,82 @@
+import * as web3 from "web3"
+import { getBytes } from "ethers"
+import { TypedDataEncoder } from "ethers/hash"
+
+import { Action, Message, Session } from "@canvas-js/interfaces"
+type Codec = { name: string; code: number; encode: (value: any) => Iterable<Uint8Array> }
+
+function dynamicAbiEncodeArgs(args: Record<string, any>): string {
+	const types: string[] = []
+	const values: any[] = []
+	for (const key in args) {
+		types.push("string")
+		values.push(key)
+		if (typeof args[key] === "string") {
+			if (args[key].match(/^0x[0-9a-fA-F]*$/)) {
+				types.push("address")
+			} else {
+				types.push("string")
+			}
+		} else if (typeof args[key] === "number") {
+			types.push("int256")
+		} else if (typeof args[key] === "boolean") {
+			types.push("boolean")
+		}
+
+		values.push(args[key])
+	}
+	return web3.eth.abi.encodeParameters(types, values)
+}
+
+export const eip712Codec: Codec = {
+	name: "eip712",
+	code: 712,
+	encode: (message: Message<Action | Session>) => {
+		// TODO: how do we pass the domain into encode?
+
+		let hashedPayload: string
+		if (message.payload.type === "session") {
+			const sessionType = [
+				{ name: "address", type: "address" }, // the address that is delegated-signing the action
+				{ name: "publicKey", type: "string" }, // the burner address that is being authorized to sign actions
+				{ name: "blockhash", type: "string" }, // may be "" if no blockhash
+				{ name: "timestamp", type: "uint256" }, // this is actually overkill at uint24 is enough, but we can revisit during code review
+				{ name: "duration", type: "uint256" },
+			]
+			const { address, publicKey, blockhash, timestamp, duration } = message.payload
+			hashedPayload = TypedDataEncoder.hash(
+				{},
+				{ Session: sessionType },
+				{ address: address.split(":")[2], publicKey, blockhash, timestamp, duration },
+			)
+		} else if (message.payload.type === "action") {
+			const actionType = [
+				{ name: "name", type: "string" },
+				{ name: "args", type: "bytes" },
+				{ name: "address", type: "address" }, // the address that is delegated-signing the action
+				// TODO: how is this passed into encode?
+				// { name: "session", type: "address" }, // the burner address that is directly signing the action. this could be omitted in the future, but let's include and then revisit during code review
+				{ name: "blockhash", type: "string" }, // may be "" if no blockhash
+				{ name: "timestamp", type: "uint256" }, // this is actually overkill at uint24 is enough, but we can revisit during code review
+				// TODO: how is this passed into encode?
+				// { name: "duration", type: "uint256" },
```

Comment:
no `duration` for action, maybe i copied it into the session type definitions by accident? i fixed this in the latest update to hicommonwealth/commonwealth#6064